src/lp.rs

```diff
@@ -0,0 +1,428 @@
+            }
+        }
+        // We passed all the tracks
+        // remain = now - started - sum(track duration)
+        // How long ago the playlist finished
+        PlayState::Finished(remain)
+    }
+
+    fn build_embed(&self) -> Box<CreateEmbed> {
+        let (lp_name, lp_id) = match &self.playlist {
+            PlaylistInfo::AlbumInfo {
+                id,
+                artist,
+                name,
+                uri,
+            } => {
+                let album_name =
+                    maybe_uri(format!("{artist} - {name}"), uri.as_ref());
+                (format!("**Album**:\n {album_name}"), id.clone())
+            }
+            PlaylistInfo::PlaylistInfo { id, name, uri } => {
+                let playlist_name = maybe_uri(name, uri.as_ref());
+                (format!("**Playlist**:\n {playlist_name}"), id.clone())
+            }
+        };
+
+        let mut embed = CreateEmbed::new().description(format!(
+            "{} - \\[{}\\]",
+            lp_name,
+            display_duration(&self.tracks.iter().map(|t| t.duration).sum())
+        ));
+        match self.now_playing() {
+            PlayState::NotStarted => {
+                embed = embed.title("Listening Party has not started yet.");
+            }
+            PlayState::Finished(_) => {
+                embed = embed.title("Listening Party has finished.");
+            }
+            PlayState::Playing { track, position } => {
+                let track_uri_ctx = track
+                    .uri
+                    .as_ref()
+                    .map(|uri| format!("{}?context={}", uri, &lp_id));
+                embed = embed
+                    .title("Listening Party in full swing! Join in!")
+                    .field(
+                        "Now playing",
+                        format!(
+                            "Track {} - {} - {} / {}",
+                            track.number,
+                            maybe_uri(&track.name, track_uri_ctx.as_ref()),
+                            display_duration(&position),
+                            display_duration(&track.duration)
+                        ),
+                        true,
+                    );
+            }
+        }
+
+        Box::new(embed)
+    }
+}
+
+// Format Duration as hh:mm:ss
+fn display_duration(duration: &chrono::Duration) -> String {
+    let allsecs = duration.num_seconds();
+    let seconds = allsecs % 60;
+    let minutes = allsecs / 60 % 60;
+    let hours = allsecs / 3600;
+    if hours > 0 {
+        format!("{}:{:0>2}:{:0>2}", hours, minutes, seconds)
+    } else {
+        format!("{:0>2}:{:0>2}", minutes, seconds)
+    }
+}
+
+// Regex to identity spotify album URIs and extract album id
+const SPOTIFY_ALBUM_RE: &str =
+    "\\bhttps://open.spotify.com/album/([a-zA-Z0-9]+)(?:\\?[a-zA-Z?=&]*)\\b";
+
+const SPOTIFY_PLAYLIST_RE: &str =
+    "\\bhttps://open.spotify.com/playlist/([a-zA-Z0-9]+)(?:\\?[a-zA-Z?=&]*)\\b";
```

Comment:
Maybe handle international urls, e.g.

```suggestion
const SPOTIFY_ALBUM_RE: &str =
    "\\bhttps://open.spotify.com(/intl-[a-z]+)?/album/([a-zA-Z0-9]+)(?:\\?[a-zA-Z?=&]*)\\b";

const SPOTIFY_PLAYLIST_RE: &str =
    "\\bhttps://open.spotify.com(/intl-[a-z]+)/playlist/([a-zA-Z0-9]+)(?:\\?[a-zA-Z?=&]*)\\b";
```

Also could use raw string literals (`r"..."`) to avoid having all those `\\`

Reply:
Fixed this and added some tests, as neither of our version worked :sweat_smile: